<div style="display: flex; align-items: center; gap: 20px; font-family: Arial, sans-serif;">
    <div style="margin-right: 30px;">
        <img src="../Sources/Logo.png" width="300" style="border-radius: 10px;"/>
    </div>
    <div style="line-height: 1.5;">
        <strong style="font-size: 18px; display: block; margin-bottom: 5px;">
            <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">
                Репетитор по информатике | Мамаев Алексей
            </a>
        </strong>
        <span style="font-size: 16px; color: #333;">
            Разбор <a href='https://euler.jakumo.org/problem/4-naibolsee-proizvedenie-palindrom' style="text-decoration: none; color: #1a73e8;">
            4-й задачи</a> проекта Эйлера
        </span>
        <span style="font-size: 16px; color: #333;">
            <br><br><br><br><br><br>ТГК: <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">@M4m4ev_Alex</a><br>
        </span>
    </div>
</div>

## Задача 4

Число-палиндром читается одинаково как слева направо, так и справа налево.

Например, наибольший палиндром, полученный произведением двух двузначных чисел, равен

$$
9009 = 91 \cdot 99
$$

Необходимо найти **наибольший палиндром**, который можно получить произведением двух трёхзначных чисел.

Ответ: 906609

## 1) Полный перебор произведений

В данном решении последовательно перебираются все возможные пары трёхзначных чисел.

Для каждой пары вычисляется произведение, после чего проверяется, является ли полученное число палиндромом. Если найденный палиндром больше текущего максимума, его значение сохраняется.

Проверка на палиндром выполняется с помощью разворота строки.

### Алгоритм

1. Создаётся переменная для хранения максимального палиндрома.
2. Последовательно перебираются все трёхзначные числа `a`.
3. Для каждого `a` перебираются все трёхзначные числа `b`.
4. Вычисляется произведение `a * b`.
5. Проверяется, совпадает ли число со своей записью в обратном порядке.
6. Если найденный палиндром больше текущего максимума, значение обновляется.
7. После завершения перебора возвращается найденный палиндром.

### Особенности

- Используется полный перебор всех пар трёхзначных чисел.
- Проверка палиндрома выполняется через разворот строки.
- Все возможные комбинации рассматриваются независимо друг от друга.
- Временная сложность составляет `O(n²)`.
- Пространственная сложность составляет `O(1)`.

### Вывод

Полный перебор позволяет получить правильный ответ без использования дополнительных математических свойств задачи. Однако при этом выполняется большое количество лишних проверок, что делает данный подход не самым эффективным.

In [16]:
from time import perf_counter # измерение времени выполнения алгоритмов

In [17]:
def get_result_1() -> int:
    m = 0
    for a in range(100, 1_000):
        for b in range(100, 1_000):
            ab = a * b
            if ab == int(str(ab)[::-1]):
                m = max(m, ab)
    return m


start = perf_counter()
print(get_result_1())
end = perf_counter()


print(f'{end - start:.5f}')

906609
0.24142


## 2) Исключение повторяющихся комбинаций

Во втором решении используется тот же полный перебор произведений, однако исключаются повторяющиеся пары чисел.

Поскольку умножение обладает переместительным свойством,

$$
a \cdot b = b \cdot a,
$$

проверять одновременно пары `(a, b)` и `(b, a)` нет необходимости.

Поэтому внутренний цикл начинается с `a + 1`, что позволяет рассматривать каждую пару только один раз.

### Алгоритм

1. Создаётся переменная для хранения максимального палиндрома.
2. Последовательно перебираются все трёхзначные числа `a`.
3. Для каждого `a` перебираются значения `b`, начиная с `a + 1`.
4. Вычисляется произведение `a * b`.
5. Проверяется, является ли произведение палиндромом.
6. Если найденный палиндром больше текущего максимума, значение обновляется.
7. После завершения перебора возвращается найденный палиндром.

### Особенности

- Исключаются повторяющиеся комбинации множителей.
- Каждая пара чисел рассматривается только один раз.
- Проверка палиндрома выполняется через разворот строки.
- Временная сложность остаётся `O(n²)`, однако количество проверок уменьшается почти в два раза.
- Пространственная сложность составляет `O(1)`.

### Вывод

Использование свойства переместительности умножения позволяет значительно сократить количество проверяемых комбинаций без изменения основной идеи алгоритма. Благодаря этому решение работает быстрее предыдущего варианта.

In [18]:
def get_result_2() -> int:
    m = 0
    for a in range(100, 1_000 - 1):
        for b in range(a + 1, 1_000):
            ab = a * b
            if ab == int(str(ab)[::-1]):
                m = max(m, ab)
    return m


start = perf_counter()
print(get_result_2())
end = perf_counter()

print(f'{end - start:.5f}')

906609
0.12144


## 3) Использование генератора

Данное решение использует ту же идею, что и предыдущий алгоритм, однако реализовано в более компактной форме.

Все произведения, являющиеся палиндромами, формируются с помощью генераторного выражения, а функция `max()` сразу возвращает наибольшее из них.

### Отличия от предыдущего решения

- Используется генераторное выражение вместо вложенных циклов.
- Максимальное значение находится встроенной функцией `max()`.
- Код становится значительно компактнее.
- Алгоритм и временная сложность остаются без изменений.

### Вывод

Использование генераторного выражения позволяет сделать решение более лаконичным, сохранив прежнюю логику работы и производительность.

In [19]:
def get_result_3() -> int:
    return max(a * b
               for a in range(100, 1_000 - 1)
               for b in range(a + 1, 1_000)
               if a * b == int(str(a * b)[::-1]))


start = perf_counter()
print(get_result_3())
end = perf_counter()

print(f'{end - start:.5f}')

906609
0.11902


## 4) Перебор в порядке убывания с ранним завершением

В данном решении множители перебираются от больших значений к меньшим. Это позволяет раньше находить крупные произведения-палиндромы и использовать их для сокращения дальнейшего поиска.

После нахождения палиндрома для текущего значения `a` внутренний цикл завершается, так как следующие значения `b` будут только меньше и дадут меньшее произведение.

Дополнительно проверяется условие:

```python
if a**2 < m:
    break
```

Если даже максимальное возможное произведение для текущего и последующих значений `a` уже меньше найденного палиндрома, дальнейший перебор не имеет смысла.

### Алгоритм

1. Значения `a` перебираются от `999` к `101`.
2. Для каждого `a` значения `b` перебираются от `a - 1` к `100`.
3. Вычисляется произведение `a * b`.
4. Если произведение является палиндромом, максимум обновляется, а внутренний цикл завершается.
5. Если `a²` меньше уже найденного максимума, внешний цикл также завершается.
6. Возвращается наибольший найденный палиндром.

### Особенности

- Перебор выполняется в порядке убывания.
- Исключаются повторяющиеся пары множителей.
- Используется раннее завершение обоих циклов.
- Количество проверок значительно уменьшается.
- Асимптотическая сложность остаётся `O(n²)`, но на практике алгоритм работает намного быстрее предыдущих решений.

### Вывод

Перебор от больших значений к меньшим позволяет быстро найти крупный палиндром и использовать его как границу для остановки поиска. Благодаря раннему завершению циклов решение выполняет значительно меньше лишних проверок.

In [20]:
def get_result_4():
    m = 0
    for a in range(999, 100, -1):
        for b in range(a - 1, 99, -1):
            ab = a * b
            if ab == int(str(ab)[::-1]):
                m = max(m, ab)
                break
        if a**2 < m:
            break
    return m
    

start = perf_counter()
print(get_result_4())
end = perf_counter()

print(f'{end - start:.5f}')

906609
0.00960


## 5) Динамическое сокращение диапазона поиска

В данном решении множители также перебираются в порядке убывания, однако границы поиска изменяются в процессе работы алгоритма.

Переменная `r` хранит текущий максимальный множитель, а `l` — нижнюю границу внутреннего перебора. После нахождения нового палиндрома значение `l` увеличивается, поэтому последующие итерации не рассматривают множители, которые уже не смогут улучшить найденный результат.

### Алгоритм

1. Начальные границы устанавливаются равными `100` и `999`.
2. Пока левая граница меньше правой, для текущего значения `r` перебираются возможные множители `tr` в порядке убывания.
3. Для каждой пары вычисляется произведение `r * tr`.
4. Если произведение является палиндромом и превышает текущий максимум:
   - максимум обновляется;
   - левая граница переносится на значение `tr`;
   - внутренний цикл завершается.
5. После обработки текущего значения правая граница уменьшается.
6. Если квадрат правой границы уже меньше найденного максимума, дальнейший поиск прекращается.
7. Возвращается наибольший найденный палиндром.

### Особенности

- Множители перебираются от больших значений к меньшим.
- Нижняя граница поиска динамически увеличивается после нахождения нового максимума.
- Используется раннее завершение внутреннего и внешнего циклов.
- Значительно сокращается количество проверяемых произведений.
- В худшем случае сложность остаётся `O(n²)`, но на практике выполняется существенно меньше операций.

### Вывод

Динамическое изменение границ позволяет постепенно сужать область поиска и не рассматривать заведомо менее перспективные пары множителей. Благодаря этому решение работает быстрее обычного полного перебора.

In [21]:
def get_result_5():
    l, r = 100, 999
    p = 0

    while l < r:
        for tr in range(r, l, -1):
            pl = r * tr
            if pl == int(str(pl)[::-1]) and pl > p:
                p = pl
                l = tr
                break
        r -= 1
        if r**2 < p:
            break
    return p
    
start = perf_counter()
print(get_result_5())
end = perf_counter()


print(f'{end - start:.5f}')

906609
0.00217


## Сравнение времени выполнения

Для сравнения производительности используется функция `timeit()`, предназначенная специально для измерения времени выполнения небольших участков кода.

Каждый алгоритм запускается **100 раз**, после чего общее время делится на количество запусков. Таким образом получается **среднее время выполнения одного запуска**.

In [24]:
from timeit import timeit

funcs = (
    get_result_1,
    get_result_2,
    get_result_3,
    get_result_4,
    get_result_5,
)

for i, func in enumerate(funcs, start=1):
    tm = timeit(lambda: func(), number=100) / 100
    print(f'Алгоритм {i}: {tm:.7f}')

Алгоритм 1: 0.2252852
Алгоритм 2: 0.1179800
Алгоритм 3: 0.1163593
Алгоритм 4: 0.0085309
Алгоритм 5: 0.0019188


## Заключение

В ходе решения задачи были рассмотрены пять различных подходов к поиску наибольшего палиндрома, полученного произведением двух трёхзначных чисел.

Первый алгоритм основан на полном переборе всех возможных пар множителей. Несмотря на простоту реализации, он выполняет большое количество повторяющихся проверок и оказался самым медленным.

Во втором решении используется свойство переместительности умножения. Благодаря исключению повторяющихся комбинаций количество проверок сокращается почти в два раза, что позволяет заметно ускорить выполнение программы.

Третий алгоритм реализует ту же идею, но с использованием генераторного выражения и встроенной функции `max()`. Код становится значительно компактнее, а производительность практически не отличается от предыдущего решения.

В четвёртом решении множители перебираются в порядке убывания. После нахождения подходящего палиндрома используется раннее завершение внутреннего цикла, а дополнительная проверка позволяет преждевременно завершить и внешний цикл. Это существенно сокращает объём вычислений.

Наиболее эффективным оказалось пятое решение. Помимо перебора в порядке убывания и раннего завершения циклов, оно динамически сужает диапазон поиска, исключая множители, которые уже не смогут привести к лучшему результату.

Среднее время выполнения алгоритмов:

- **Алгоритм 1:** `0.2252852` с
- **Алгоритм 2:** `0.1179800` с
- **Алгоритм 3:** `0.1163593` с
- **Алгоритм 4:** `0.0085309` с
- **Алгоритм 5:** `0.0019188` с

Полученные результаты показывают, что основное ускорение достигается не за счёт изменения синтаксиса, а благодаря сокращению количества проверяемых комбинаций. Использование раннего завершения циклов и динамического уменьшения области поиска позволило сделать пятый алгоритм более чем в **117 раз быстрее** полного перебора.